src/vm/loaderallocator.cpp

```diff
@@ -2046,4 +2053,126 @@ BOOL LoaderAllocator::InsertComInteropData(MethodTable* pMT, InteropMethodTableD
+    }
+    CONTRACTL_END;
+    if (libraryName)
+    {
+        m_UnmanagedCache.InsertEntry(libraryName, hMod);
+    }
+    return ;
+}
+
+HMODULE LoaderAllocator::FindUnmanagedImageInCache(LPCWSTR libraryName)
+{
+    CONTRACT(HMODULE)
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(libraryName,NULL_OK));
+        POSTCONDITION(CheckPointer(RETVAL,NULL_OK));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACT_END;
+    if(libraryName == NULL) RETURN NULL;
+
+    RETURN m_UnmanagedCache.LookupEntry(libraryName);
+}
+
+/* static */
+BOOL UnmanagedImageCache::CompareBindingSpec(UPTR spec1, UPTR spec2)
+{
+    WRAPPER_NO_CONTRACT;
+
+    LPCWSTR libraryName = (LPCWSTR) (spec1 << 1);
+    ImageEntry* pEntry2 = (ImageEntry*) spec2;
+
+    return wcscmp(libraryName, pEntry2->libraryName) == 0;
+}
+
+HMODULE UnmanagedImageCache::LookupEntry(LPCWSTR libraryName)
+{
+    CONTRACT (HMODULE)
+    {
+        INSTANCE_CHECK;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        POSTCONDITION(CheckPointer(RETVAL, NULL_OK));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACT_END
+
+    DWORD hashValue = HashString(libraryName);
+
+    LPVOID pResult = m_Table.LookupValue(hashValue, (LPVOID)libraryName);
+    if(pResult == (LPVOID) INVALIDENTRY)
+        RETURN NULL;
+    else
+        RETURN ((ImageEntry*)pResult)->hMod;
+}
+
+void UnmanagedImageCache::InsertEntry(LPCWSTR libraryName, HMODULE hMod)
+{
+    CONTRACTL
+    {
+        INSTANCE_CHECK;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END
+
+    HMODULE hCachedMod = LookupEntry(libraryName);
+    if(hCachedMod == NULL)
+    {
+        LoaderAllocator::CacheLockHolder lh(m_pLoaderAllocator);
+
+        hCachedMod = LookupEntry(libraryName);
+        if (hCachedMod == NULL) 
+        {
+            AllocMemTracker amTracker;
+            AllocMemTracker *pamTracker = &amTracker;
+
+            ImageEntry* pEntry = (ImageEntry*) pamTracker->Track( m_pLoaderAllocator->GetLowFrequencyHeap()->AllocMem(S_SIZE_T(sizeof(ImageEntry))) );
+
+            size_t len = wcslen(libraryName) + 1;
+            LPWSTR temp = (LPWSTR)pamTracker->Track( m_pLoaderAllocator->GetLowFrequencyHeap()->AllocMem(S_SIZE_T (len * 2)) );
+            wcscpy_s(temp, len, libraryName);
+            pEntry->libraryName = temp;
+            pEntry->hMod = hMod;
+            DWORD hashValue = HashString(libraryName);
+            m_Table.InsertValue(hashValue, pEntry);
+
+            pamTracker->SuppressRelease();
+        }
+        // lh goes out of scope here
+    }
+#ifdef _DEBUG
+    else 
+    {
+        _ASSERTE(hMod == hCachedMod);
+    }
+#endif
+}
+
+void UnmanagedImageCache::FreeImages()
+{
+    for (PtrHashMap::PtrIterator it = m_Table.begin(); !it.end(); ++it)
+    {
+        FreeLibrary(((ImageEntry*)it.GetValue())->hMod);
```

Comment:
I do not think we have freed the unmanaged libraries even with AppDomains. I know that the security experts have a very strong reservations about unloading unmanaged libraries by default (or make it otherwise easy) Most unmanaged libraries are not designed to be ever unloaded.

Reply:
cc @GrabYourPitchforks

Reply:
Agreed. I thought that this kind of management was the some what the point of the `NativeLibrary` API. If users want this kind of unload behavior they should control load and unload - at least that was what I thought was one of the reasons for the API.

Reply:
It seems kind of logical to me to unload everything managed and unmanaged when you unload a context. I understand the concerns about possible issues with unloading unmanaged libraries though.